drivers/spi/spi_ambiq_spis.c

```diff
@@ -0,0 +1,382 @@
+	struct spi_ambiq_data *data = dev->data;
+
+	am_hal_ios_interrupt_status_get(data->ios_handler, false, &ui32Status);
+	am_hal_ios_interrupt_clear(data->ios_handler, ui32Status);
+	if (ui32Status & AM_HAL_IOS_INT_XCMPWR) {
+		k_sem_give(&data->spim_wrcmp_sem);
+	}
+}
+
+static int spi_config(const struct device *dev, const struct spi_config *config)
+{
+	struct spi_ambiq_data *data = dev->data;
+	struct spi_context *ctx = &(data->ctx);
+	int ret = 0;
+
+	data->ios_cfg.ui32InterfaceSelect = AM_HAL_IOS_USE_SPI;
+
+	if (spi_context_configured(ctx, config)) {
+		/* Already configured. No need to do it again. */
+		return 0;
+	}
+
+	if (SPI_WORD_SIZE_GET(config->operation) != AMBIQ_SPIS_WORD_SIZE) {
+		LOG_ERR("Word size must be %d", AMBIQ_SPIS_WORD_SIZE);
+		return -ENOTSUP;
+	}
+
+	if ((config->operation & SPI_LINES_MASK) != SPI_LINES_SINGLE) {
+		LOG_ERR("Only supports single mode");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_LOCK_ON) {
+		LOG_ERR("Lock On not supported");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_TRANSFER_LSB) {
+		LOG_ERR("LSB first not supported");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_MODE_CPOL) {
+		if (config->operation & SPI_MODE_CPHA) {
+			data->ios_cfg.ui32InterfaceSelect |= AM_HAL_IOS_SPIMODE_3;
+		} else {
+			data->ios_cfg.ui32InterfaceSelect |= AM_HAL_IOS_SPIMODE_2;
+		}
+	} else {
+		if (config->operation & SPI_MODE_CPHA) {
+			data->ios_cfg.ui32InterfaceSelect |= AM_HAL_IOS_SPIMODE_1;
+		} else {
+			data->ios_cfg.ui32InterfaceSelect |= AM_HAL_IOS_SPIMODE_0;
+		}
+	}
+
+	if (config->operation & SPI_OP_MODE_MASTER) {
+		LOG_ERR("Controller mode not supported");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_MODE_LOOP) {
+		LOG_ERR("Loopback mode not supported");
+		return -ENOTSUP;
+	}
+
+	if (spi_cs_is_gpio(config)) {
+		LOG_ERR("CS control via GPIO is not supported");
+		return -EINVAL;
+	}
+
+	/* Eliminate the "read-only" section, so an external host can use the
+	 * entire "direct write" section.
+	 */
+	data->ios_cfg.ui32ROBase = AMBIQ_SPIS_FIFO_BASE;
+	/* Making the "FIFO" section as big as possible. */
+	data->ios_cfg.ui32FIFOBase = AMBIQ_SPIS_FIFO_BASE;
+	/* We don't need any RAM space, so extend the FIFO all the way to the end
+	 * of the LRAM.
+	 */
+	data->ios_cfg.ui32RAMBase = AMBIQ_SPIS_FIFO_END;
+	/* FIFO Threshold - set to half the size */
+	data->ios_cfg.ui32FIFOThreshold = AMBIQ_SPIS_FIFO_LENGTH >> 1;
+
+	data->ios_cfg.pui8SRAMBuffer = ambiq_spis_sram_buffer,
+	data->ios_cfg.ui32SRAMBufferCap = AMBIQ_SPIS_TX_BUFSIZE_MAX,
+
+	ctx->config = config;
+
+	ret = am_hal_ios_configure(data->ios_handler, &data->ios_cfg);
+
+	return ret;
+}
+
+static int spi_ambiq_xfer(const struct device *dev, const struct spi_config *config)
+{
+	const struct spi_ambiq_config *cfg = dev->config;
+	struct spi_ambiq_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int ret = 0;
+	uint32_t chunk, num_written, num_read = 0;
+
+	while (1) {
```

Comment:
The logic in this loop does not seem right. 

A SPI transaction is not "First I send all, then I read all". If you send 1mb, but the controller has an rx internal buffer of 1kb... you'll get into trouble. I don't know what this LRAM is made of, but I doubt it's limitless. 

Also, is calling am_hal_ios_fifo_write() with a NULL as a 2nd parameter  going to work? Because it can be NULL. spi_buf->len can be > 0 with spi_buf->tx_buf being NULL. Meaning it's up to the driver (or HAL if it's clever enough) to send dummy bytes instead. And it can be the same in rx so the memcpy below is definitely going to blow up.

Have you tried running tests/drivers/spi/spi_loopback ?

Reply:
Thank you for the comments, let me briefly introduce how our SPIS works.
Ambiq's SPIS (we called it IOS which is I/O Slave) is very specific designed, basically it contains 1KBytes share memory (Local RAM) and the SPI interface, it cannot "send" data out on its own, all operations have to be initiated by the host, so if the slave need to "send" 1mb to the host, we first trigger a GPIO interrupt to the host and then fill as much data to the FIFO, and once the host received the GPIO intrerrupt, it will first read the FIFO counter register FIFOCTR of the slave to know how much data it need to read at that time, after it read the data, the slave will start another iteration to "send" the rest data. In short there must be some kind of handshake before any data transfer.
And Apollo3 SPIS hardware does not support dummy bytes send, that's why I added spi_context_tx_buf_on and spi_context_rx_buf_on check there, NULL buffers should not be passed to HAL.
As our SPIS is actually totally different design compared with SPIM, I don't think it can run the spi_loopback test, we tested the driver with our local testcases.

Reply:
I missed the fact it is a spi**S** controller, nevermind. (as well as I mixed up tx/rx_buf_on with tx/rx_on  -_-' )

spi_controller_peripheral should be the test to use. (if you have 2 SPI controllers - once acting as the host, the other one as the peripheral - on one board it's quite trivial to set up)

You can't avoid managing the dummy bytes in the driver: if the tx_buf is NULL but tx_len is > 0, it means it should tx dummy bytes of tx_len amount. Or else user code won't be easily portable on your driver. You could allocate a small static array of dummy bytes, and you loop on writing these relevantly to the necessary amount (if the dummy bytes array has a length of 16, if the user needs to tx 17 dummy bytes: it loops twice, once with 16, the next time with 1. For instance.)

Same on rx side, don't you need to "consume" all bytes that have been sent to you? Even the user does not provide any buffer for it I mean. (I does not have to be dummy bytes, it can be any bytes)

the spi buf set is of scatter/gather type, so it can be a complicated setup with various tx/rx, sometimes with no actual buffer but just a length there. That's why I find it weird that you do not loop until it has went through all the set (basically the goto end; in the rx block).

Reply:
I added dummy bytes read/write support, please help to review again, thank you!